majortom_gateway/gateway_api.py

```diff
@@ -0,0 +1,375 @@
+            try:
+                return await self.connect()
+            except (OSError, asyncio.streams.IncompleteReadError, websockets.ConnectionClosed) as e:
+                self.websocket = None
+                logger.warning("Connection error encountered, retrying in 5 seconds ({})".format(e))
+                await asyncio.sleep(5)
+            except Exception as e:
+                logger.error("Unhandled {} in `connect_with_retries`".format(e.__class__.__name__))
+                raise e
+
+    async def handle_message(self, json_data):
+        message = json.loads(json_data)
+        message_type = message["type"]
+        logger.debug("From Major Tom: {}".format(message))
+        if message_type == "command":
+            command = Command(message["command"])
+            if self.command_callback is not None:
+                """
+                TODO: Track the task and ensure it completes without errors
+                reference: https://medium.com/@yeraydiazdiaz/asyncio-coroutine-patterns-errors-and-cancellation-3bb422e961ff
+                """
+                asyncio.ensure_future(self.command_callback(command, self))
+            else:
+                asyncio.ensure_future(self.fail_command(
+                    command.id, errors=["No command callback implemented"]))
+        elif message_type == "cancel":
+            if self.cancel_callback is not None:
+                """
+                TODO: Track the task and ensure it completes without errors
+                reference: https://medium.com/@yeraydiazdiaz/asyncio-coroutine-patterns-errors-and-cancellation-3bb422e961ff
+                """
+                asyncio.ensure_future(self.cancel_callback(message["command"]["id"], self))
+            else:
+                asyncio.ensure_future(self.transmit_events(events=[{
+                    "system": None,
+                    "type": "Command Cancellation Failed",
+                    "command_id": message["command"]["id"],
+                    "level": "warning",
+                    "message": "No cancel callback registered. Unable to cancel command."
+                }]))
+        elif message_type == "error":
+            logger.error("Error from Major Tom: {}".format(message["error"]))
+            if self.error_callback is not None:
+                asyncio.ensure_future(self.error_callback(message))
+        elif message_type == "rate_limit":
+            logger.error("Rate limit from Major Tom: {}".format(message["rate_limit"]))
+            if self.rate_limit_callback is not None:
+                asyncio.ensure_future(self.rate_limit_callback(message))
+        elif message_type == "hello":
+            logger.info("Major Tom says hello: {}".format(message))
+        else:
+            logger.warning("Unknown message type {} received from Major Tom: {}".format(
+                message_type, message))
+
+    async def empty_queue(self):
+        while len(self.queued_payloads) > 0 and self.websocket:
```

Comment:
This should sleep an appropriate amount when emptying the buffer to avoid hitting rate limits.

Reply:
Also a TODO I think